pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+	}
+	switch node, event := nodeEvent.Node, nodeEvent.Event; event {
+	case memberlist.NodeJoin:
+		// when node join cluster, all matched externalIPPools consistentHash should changed
+		node, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			klog.ErrorS(err, "Processing clusterNode events failed")
+			return
+		}
+		count := 0
+		for _, eip := range eips {
+			if c.matchNodeSelector(node, eip) {
+				c.queue.Add(eip.Name)
+				klog.V(5).InfoS("Node join event", "nodeName", node.String(), "eip", eip.Name)
+				count++
+			}
+		}
+		klog.InfoS("Node event: join node", "nodeName", node.Name, "eipNum", count)
+	case memberlist.NodeLeave:
+		// when node leave cluster, the node may failed or deleted, all externalIPPools consistentHash maybe changed
+		for _, eip := range eips {
+			c.queue.Add(eip.Name)
+			klog.V(5).InfoS("Node leave event", "nodeName", node.String(), "eip", eip.Name)
+		}
+		klog.InfoS("Node event: leave node", "nodeName", node.Name, "eipNum", len(eips))
+	case memberlist.NodeUpdate:
+		klog.InfoS("Node event: update node", "nodeName", node.Name)
+	default:
+		klog.InfoS("Node event: unknown", "nodeName", node.Name, "event", event)
+	}
+}
+
+// aliveNodeList return alive nodeName list in cluster
+func (c *Cluster) aliveNodeList() sets.String {
+	aliveMembers := c.mList.Members()
+	nodes := make([]string, len(aliveMembers))
+	for i, node := range aliveMembers {
+		nodes[i] = node.Name
+	}
+	return sets.NewString(nodes...)
+}
+
+func (c *Cluster) shouldSelect(consistentHash *consistenthash.Map, name string) bool {
+	myNode := c.nodeName
+	hitted := hitNodeByConsistentHash(consistentHash, name, myNode)
+	klog.InfoS("Assign egress owner for local node", "egressName", name, "localNode", myNode, "shouldSelected", hitted)
+	return hitted
+}
+
+// ShouldSelectEgress only node which joined cluster need to hold egress
+func (c *Cluster) ShouldSelectEgress(egress *v1alpha2.Egress) bool {
+	eipName := egress.Spec.ExternalIPPool
+	if eipName == "" {
+		return false
+	}
+	// local node must matches externalIPPool
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		klog.ErrorS(err, "Select node for egress failed", "eip", eipName)
+		return false
+	}
+	localNode, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		klog.ErrorS(err, "Select node for egress failed", "nodeName", c.nodeName)
+		return false
+	}
+	if !c.matchNodeSelector(localNode, eip) {
+		klog.ErrorS(err, "Select node for egress eip not match", "nodeName", c.nodeName)
+		return false
+	}
+	// local in the externalIPPool, will participate failover, then check if local node should be selected
+	c.consistentHashRWMutex.RLock()
+	defer c.consistentHashRWMutex.RUnlock()
+	egressName := egress.Name
+	consistentHash, ok := c.consistentHashMap[eipName]
+	if !ok {
+		klog.InfoS("Local node consistentHashMap has not synced", "eip", eipName)
+		return false
+	}
+	return c.shouldSelect(consistentHash, egressName)
```

Comment:
I think it's a bit overkill to have this calling stack: `c.shouldSelect` -> `hitNodeByConsistentHash` -> `consistentHash.Get(name) == myNode`. 
Since we already fetched the `consistentHash` here and the method is already named `ShouldSelectEgress`, why don't just call `consistentHash.Get(name) == myNode` directly?

Reply:
done